controllers/topology/internal/extensions/patches/inline/json_patch_generator.go

```diff
@@ -0,0 +1,293 @@
+	// they cannot be directly consumed as byte arrays.
+	data, err := calculateTemplateData(variables)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to calculate template data")
+	}
+
+	// Render the template.
+	var buf bytes.Buffer
+	if err := tpl.Execute(&buf, data); err != nil {
+		return nil, errors.Wrapf(err, "failed to render template: %q", valueTemplate)
+	}
+
+	// Unmarshal the rendered template.
+	// NOTE: The YAML library is used for unmarshalling, to be able to handle YAML and JSON.
+	value := apiextensionsv1.JSON{}
+	if err := yaml.Unmarshal(buf.Bytes(), &value); err != nil {
+		return nil, errors.Wrapf(err, "failed to unmarshal rendered template: %q", buf.String())
+	}
+
+	return &value, nil
+}
+
+// calculateTemplateData calculates data for the template, by converting
+// a flat variables map to a nested map and converting the variables to
+// their Go types.
+// Example:
+// * Input:
+//   map[string]apiextensionsv1.JSON{
+//     "builtin.cluster.name": {Raw: []byte(`"cluster-name"`},
+//     "integerVariable": {Raw: []byte("4")},
+//     "numberVariable": {Raw: []byte("2.5")},
+//     "booleanVariable": {Raw: []byte("true")},
+//   }
+// * Output:
+//   map[string]interface{}{
+//     "builtin": map[string]interface{}{
+//       "cluster": map[string]interface{}{
+//         "name": <string>"cluster-name"
+//       }
+//     },
+//     "integerVariable": <float64>4,
+//     "numberVariable": <float64>2.5,
+//     "booleanVariable": <bool>true,
+//   }
+func calculateTemplateData(variables map[string]apiextensionsv1.JSON) (map[string]interface{}, error) {
+	res := map[string]interface{}{}
+
+	// Iterate through all variables.
+	for name, value := range variables {
+		// Split the variable name in segments
+		nameSegments := strings.Split(name, ".")
+
+		// Iterate through the segments and created nested maps,
+		// except for the latest segment.
+		currentMap := res
+		for i := 0; i < len(nameSegments)-1; i++ {
+			segment := nameSegments[i]
+
+			// Created the nested map if it doesn't already exist.
+			if _, ok := currentMap[segment]; !ok {
+				currentMap[segment] = map[string]interface{}{}
+			}
+			currentMap = currentMap[segment].(map[string]interface{})
+		}
+
+		// Convert the variable to its Go type.
+		var v interface{}
+		if err := json.Unmarshal(value.Raw, &v); err != nil {
+			return nil, errors.Wrapf(err, "failed to convert variable %q (value: %q) to Go type", name, string(value.Raw))
+		}
+
+		// Add the variable to the leaf map.
+		currentMap[nameSegments[len(nameSegments)-1]] = v
+	}
+
+	return res, nil
+}
+
+// mergeVariables merges variables.
+// NOTE: In case a variable exists in both maps, the variable from the first map is preserved.
+func mergeVariables(a, b map[string]apiextensionsv1.JSON) map[string]apiextensionsv1.JSON {
+	m := make(map[string]apiextensionsv1.JSON)
+	for k, v := range b {
+		m[k] = v
+	}
+	for k, v := range a {
+		m[k] = v
+	}
+	return m
```

Comment:
I'd have expected the later arguments to overwrite the preceding ones here, any reason we've chosen this way?

Reply:
I aligned to what we did in desired state: https://github.com/kubernetes-sigs/cluster-api/blob/bbefd811f2424cbb3f31d336f17f58f235061b3c/controllers/topology/desired_state.go#L651 and tried to keep it consistent to avoid confusion. But I'm fine with inverting it.

Reply:
After a bit of Google and research in the k8s.io codebase (labels.Merge) I agree. Let's flip the order